pkg/uefi/nvram.go

```diff
@@ -0,0 +1,458 @@
+
+// NVarEntrySignature value for 'NVAR' signature
+const NVarEntrySignature uint32 = 0x5241564E
+
+// NVarHeader represents an NVAR entry header
+type NVarHeader struct {
+	Signature  uint32 `json:"-"`
+	Size       uint16
+	Next       [3]uint8 `json:"-"`
+	Attributes NVarAttribute
+}
+
+// NVarEntryType represent the computed type of a NVAR entry
+type NVarEntryType uint8
+
+// Types
+const (
+	InvalidNVarEntry NVarEntryType = iota
+	InvalidLinkNVarEntry
+	LinkNVarEntry
+	DataNVarEntry
+	FullNVarEntry
+)
+
+var nVarEntryTypeName = map[NVarEntryType]string{
+	InvalidNVarEntry:     "Invalid",
+	InvalidLinkNVarEntry: "Invalid link",
+	LinkNVarEntry:        "Link",
+	DataNVarEntry:        "Data",
+	FullNVarEntry:        "Full",
+}
+
+func (t NVarEntryType) String() string {
+	if s, ok := nVarEntryTypeName[t]; ok {
+		return s
+	}
+	return "UNKNOWN"
+}
+
+// NVarExtAttribute represent extended attributes
+type NVarExtAttribute uint8
+
+// Extended attributes values
+const (
+	NVarEntryExtChecksum    NVarExtAttribute = 0x01
+	NVarEntryExtAuthWrite   NVarExtAttribute = 0x10
+	NVarEntryExtTimeBased   NVarExtAttribute = 0x20
+	NVarEntryExtUnknownMask NVarExtAttribute = 0xCE
+)
+
+// NVar represent an NVAR entry
+type NVar struct {
+	Header    NVarHeader
+	GUID      guid.GUID
+	GUIDIndex *uint8 `json:",omitempty"`
+	Name      string
+
+	NVarStore *NVarStore `json:",omitempty"`
+
+	//Decoded data
+	Type       NVarEntryType
+	Offset     uint64
+	NextOffset uint64
+
+	//Extended Header
+	ExtAttributes               *NVarExtAttribute `json:",omitempty"`
+	Checksum                    *uint8            `json:",omitempty"`
+	ExpectedChecksum            *uint8            `json:",omitempty"`
+	TimeStamp                   *uint64           `json:",omitempty"`
+	Hash                        []byte            `json:",omitempty"`
+	UnknownExtendedHeaderFormat bool              `json:",omitempty"`
+
+	//Metadata for extraction and recovery
+	buf        []byte
+	DataOffset int64
+	ExtOffset  int64 `json:",omitempty"`
+}
+
+// NVarStore represent an NVAR store
+type NVarStore struct {
+	buf []byte
+
+	Entries []*NVar
+
+	GUIDStoreLen uint8 `json:",omitempty"`
+}
+
+// Buf returns the buffer.
+// Used mostly for things interacting with the Firmware interface.
+func (s *NVarStore) Buf() []byte {
+	return s.buf
+}
+
+// SetBuf sets the buffer.
+// Used mostly for things interacting with the Firmware interface.
+func (s *NVarStore) SetBuf(buf []byte) {
+	s.buf = buf
+}
+
+// Apply calls the visitor on the NVarStore.
+func (s *NVarStore) Apply(v Visitor) error {
+	return v.Visit(s)
+}
+
+// ApplyChildren calls the visitor on each child node of NVarStore.
+func (s *NVarStore) ApplyChildren(v Visitor) error {
+	return nil
+}
+
+func (s *NVarStore) getGUIDFromStore(i uint8) guid.GUID {
+	var GUID guid.GUID
+	if s.GUIDStoreLen < i+1 {
+		s.GUIDStoreLen = i + 1
+	}
+	r := bytes.NewReader(s.buf)
+	if _, err := r.Seek(-int64(binary.Size(GUID))*int64(i+1), io.SeekEnd); err != nil {
```

Comment:
There are some cases, such as extract where the standard behaviour is to not save as a binary what can be stored into the JSON. This means that the buffer won't exist when you first call ParseDir. I strongly suggest instead of relying on the presence of the buffer, directly parse all of the GUIDs at the point of parsing the NVAR store instead of relying on the existence of a buffer.

Reply:
In this case, this function would instead lookup the slice of GUIDs directly rather than seeking from the buffer.

Reply:
Yes seen that when implementing `ParseDir`.
The thing to take care is that GUID are stored in reverse order and we only know how much GUID are stored when the parsing is done.
I'm going rework that.

Reply:
Thank you!